File: two_pointers/concatenate.py

def concatenate(arr1, arr2):
    first_pointer = 0
    second_pointer = 0
    result = []

    while first_pointer < len(arr1) and second_pointer < len(arr2):
        if arr1[first_pointer] <= arr2[second_pointer]:
            result.append(arr1[first_pointer])
            first_pointer += 1
            continue

        if arr1[first_pointer] > arr2[second_pointer]:
            result.append(arr2[second_pointer])
            second_pointer += 1
            continue

    result.extend(arr1[first_pointer:])
    result.extend(arr2[second_pointer:])
    
    return result


result = concatenate([1, 3, 5, 7], [2, 4, 6, 8]) # [1, 2, 3, 4, 5, 6, 7, 8]
result2 = concatenate([1, 3, 3, 7], [2, 3, 3, 8]) # [1, 2, 3, 3, 3, 3, 7, 8]

print(result)
print(result2)

File: two_pointers/concatenate.js


function concatenate(arr1, arr2) {
    const result = [];
    let first_pointer = 0;
    let second_pointer = 0;

    while (first_pointer < arr1.length && second_pointer < arr2.length) {
        if (arr1[first_pointer] <= arr2[second_pointer]) {
            result.push(arr1[first_pointer])
            first_pointer++
        } else {
            result.push(arr2[second_pointer])
            second_pointer++
        }
    }

    result.push(...arr1.slice(first_pointer))
    result.push(...arr2.slice(second_pointer))

    return result
}


const result = concatenate([1, 3, 5, 7], [2, 4, 6, 8]) // [1, 2, 3, 4, 5, 6, 7, 8]
const result2 = concatenate([1, 3, 3, 7], [2, 3, 3, 8]) // [1, 2, 3, 3, 3, 3, 7, 8]

console.log(result);
console.log(result2);